### ISSUE TITLE
Output to SE not working properly on CentOS7

### ISSUE THREAD

**Issue by kschweiger** [grid-control/grid-control#73]
Initializing a CMSSW environment breaks gfal-tools on CentOS7 (e.g. lxplus7). This leads to failing jobs in GC when staging out to a SE because gfal-ls is always used to verify that the output was copied to the destination on the SE

-----------

The general problem can be reproduced by running gfal-copy and checking the output on lxlpus7:

Before initializing CMSSW:
```bash
$ gfal-copy
usage: gfal-copy [-h] [-V] [-v] [-D DEFINITION] [-t TIMEOUT] [-E CERT]
                 [--key KEY] [-4] [-6] [-C CLIENT_INFO] [-f] [-p]
                 [-n NBSTREAMS] [--tcp-buffersize TCP_BUFFERSIZE]
                 [-s SRC_SPACETOKEN] [-S DST_SPACETOKEN] [-T TRANSFER_TIMEOUT]
                 [-K CHECKSUM] [--copy-mode {pull,push,streamed,}]
                 [--checksum-mode {source,target,both}]
                 [--from-file FROM_FILE] [--just-copy] [-r]
                 [--abort-on-failure] [--dry-run]
                 [src] dst [dst ...]
gfal-copy: error: too few arguments
```

After initializing CMSSW:
```bash
$ gfal-copy
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
ImportError: No module named site
```

This can be "fixed" by unsetting `LD_LIBRARY_PATH` and `PYTHONPATH` for the execution of `gfal-copy`:

```bash
$ LD_LIBRARY_PATH='' PYTHONPATH='' gfal-copy
usage: gfal-copy [-h] [-V] [-v] [-D DEFINITION] [-t TIMEOUT] [-E CERT]
                 [--key KEY] [-4] [-6] [-C CLIENT_INFO] [-f] [-p]
                 [-n NBSTREAMS] [--tcp-buffersize TCP_BUFFERSIZE]
                 [-s SRC_SPACETOKEN] [-S DST_SPACETOKEN] [-T TRANSFER_TIMEOUT]
                 [-K CHECKSUM] [--copy-mode {pull,push,streamed,}]
                 [--checksum-mode {source,target,both}]
                 [--from-file FROM_FILE] [--just-copy] [-r]
                 [--abort-on-failure] [--dry-run]
                 [src] dst [dst ...]
gfal-copy: error: too few arguments
```
--------------------
Ultimately this should be fixed by the CMSSW people but for a temporary solution the unsetting method can be used inside GC. [See this commit](https://github.com/kschweiger/grid-control/commit/cafa59041f3920b40160bfac8c6ed5c4cbb021f3)

I guess this should also be added somewhere [here](https://github.com/grid-control/grid-control/blob/master/packages/grid_control/share/gc-run.lib#L507) but I did not observe problems with this on T3@PSI where I run jobs (and output to the local SE). Probably because of the built in fallback.

**Comment by jolange** [comment]
I agree that this should be fixed within the CMSSW setup, because with implementing this "hack" in GC, other things might break. (If the environment variables are set, that's normally for a reason.)

So I guess for now having this in your private branch is the best solution. If there is any demand, we can make a "cms_el7" branch here for others to use. But I would not want to merge this into the master/testing branches.

**Comment by kschweiger** [comment]
Sure. I mainly wanted to bring this to the attention of the maintainers and users. Since it seems like clusters transition away from slc6 so more people might run into this.

**Comment by jolange** [comment]
Yes, thanks for that! I will also leave this issue open.

I also tested this on DESY/NAF EL7 machines and see the same problem (though it fails with a different module missing). But nobody complained yet, because I guess most people here are not staging out to dCache but to a "local" file system.

But we will certainly have to figure this out (even independent of grid-control).

**Comment by jolange** [comment]
The most useful discussion I found is cms-sw/cmssw#26462 (same symptoms e.g. in https://github.com/cms-sw/cmssw/issues/26462#issuecomment-483396388).

Stefano Belforte suggests ``(eval `scram unsetenv -sh`; gfal-ls)`` and says he gave up on having CMS and Grid software coexist...
Maybe we should also start doing this for all usage of grid tools in our scripts. @FredStober what do you think?

**Comment by jolange** [comment]
Ok, while trying to implement my solution, I realized it's already implemented "properly". I wanted to use `unsetenv` after all CMSSW related stuff is done, i.e. like this:

```
# [copy input files]
# [scram runtime setup]
cmsRun ....
# [epilog scripts]
eval $(scram unsetenv -sh)
# [copy stuff with gfal etc.]
```
But that's not necessary, because the CMSSW setup happens in a script that is not sourced. So afterwards, when the copying happens, the environment is not touched by the CMSSW setup inside the script that was used to call `cmsRun`.
I also verified this on EL7: With the current GC version, copying to dCache from CMSSW jobs works.

This means that that you probably have set up CMSSW *before* GC does it inside the job. I can reproduce this starting `grid-control` after sourcing CMSSW in the same terminal and submitting jobs to `host`.

Do you start grid-control after sourcing CMSSW? That's not a problem normally (I do the same). But maybe you forward your local environment to the jobs? Which batch system are you using? In HTCondor you can do that with `getenv = True` (which is generally not recommended to do).

**Comment by kschweiger** [comment]
Thanks for looking into this! 

I submit from a shell that has CMSSW sourced. We recently switched to SLURM so I'm not too familiar with the details yet. I'll look into it.

**Comment by FredStober** [comment]
https://slurm.schedmd.com/faq.html - #11 slurm will forward your current environment - at least for srun. It's not described what it does for sbatch.

I guess you can just specify some "--export", so you don't have to jump through hoops before running grid-control. However that needs a little config option in the slurm backend.

My suggestion would be to add a site specific config file to grid-control/config and setting that option in such a way that sbatch is called with
`--export=ALL,LD_LIBRARY_PATH='',PYTHONPATH=''`
OR
`--export=NONE`
